### PR TITLE
Revit toolkit MEP Converts - Issue fixes

### LINKED ISSUE #840: Revit_Toolkit: Local rotation of Duct members
#### Description:

OrientationAngle of Duct members is currently hardcoded and should be investigated and fixed. This property refers to the plan view rotation angle, not the slope of the duct. 

#### Steps to reproduce:

Pull any Duct object from a Revit file into a GH canvas and view the OrientationAngle value. 

#### Expected behaviour:

OrientationAngle should be accurately captured.

### PR DESCRIPTION
**Pull request by AndreBalasian**
### Issues addressed by this PR

Closes
#840 
#841 
#842 
#844 
#845 
#846 

### Test files
[Revit and GS files](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue842%2DMEPConvertIssueFixes&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4)

### Changelog
Fixed the orientation angle to display the correct angle
Fixed descriptions 
Duct shapes derived from connectors in Revit 2018
Switch in DuctSectionProfile changed for compliance
Fixed unit conversion issue
Fixed an MEP SectionProperty issue

### CODE REVIEW

File: Revit_Core_Engine/Convert/Environment/FromRevit/Pipe.cs [hunk 1]
```diff
@@ -41,7 +40,7 @@ public static partial class Convert
         [Input("revitPipe", "Revit pipe to be converted.")]
         [Input("settings", "Revit settings.")]
         [Input("refObjects", "Referenced objects.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Convert/Environment/FromRevit/Pipe.cs:42]: I would consider rewriting these 2 descriptions - they will reappear in multiple places so might be worth. What about:

```suggestion
        [Input("settings", "Revit adapter settings.")]
        [Input("refObjects", "A collection of objects processed in the current adapter action, stored to avoid processing the same object more than once.")]
```

File: Revit_Core_Engine/Query/DuctSectionProperty.cs [hunk 2]
```diff
@@ -40,21 +36,19 @@ public static partial class Query
...
         /***************************************************/
 
-        [Description("Duct section property converted from a Revit duct.")]
-        [Input("revitDuct", "Revit duct to be conerted into a section property.")]
+        [Description("Convert a Revit duct into a BHoM duct section property.")]
+        [Input("revitDuct", "Revit duct to be converted into a BHoM section property.")]
         [Input("settings", "Revit settings.")]
         [Input("refObjects", "Referenced objects.")]
-        [Output("DuctSectionProperty", "BHoM duct section property.")]
+        [Output("sectionProperty", "BHoM duct section property converted from a Revit duct.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/DuctSectionProperty.cs:43]: It is query/extraction, not conversion imo.

File: Revit_Core_Engine/Query/DuctSectionProfile.cs [hunk 3]
```diff
@@ -36,12 +36,12 @@ public static partial class Query
...
-
-        [Description("Section profile of round and rectangular ducts converted from a Revit duct.")]
-        [Input("revitDuct", "Revit duct to be converted into a section profile.")]
+        
+        [Description("Convert a Revit duct into a BHoM duct section profile.")]
+        [Input("revitDuct", "Revit duct to be converted into a BHoM section profile.")]
         [Input("settings", "Revit settings.")]
         [Input("refObjects", "Referenced objects.")]
-        [Output("profile", "Converted BHoM duct section profile converted.")]
+        [Output("profile", "BHoM section profile for a duct converted from a Revit duct.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/DuctSectionProfile.cs:44]: It is not conversion, rather extraction.

File: Revit_Core_Engine/Query/DuctShape.cs [hunk 4]
```diff
@@ -54,15 +47,28 @@ public static Autodesk.Revit.DB.ConnectorProfileType DuctShape(this Autodesk.Rev
                 return Autodesk.Revit.DB.ConnectorProfileType.Invalid;
             }
 
-            string ductFamilyName = duct?.DuctType?.FamilyName;
+            ConnectorProfileType connectorShape = ConnectorProfileType.Invalid;
 
+#if REVIT2018
+            // Get the duct connector shape in Revit 2018 by extracting it from one of the duct connectors
+            var connectors = duct.ConnectorManager.Connectors;
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/DuctShape.cs:54]: I believe this needs to take the first `End` type connector instead of any connector.

File: Revit_Core_Engine/Query/DuctShape.cs [hunk 5]
```diff
@@ -54,15 +47,28 @@ public static Autodesk.Revit.DB.ConnectorProfileType DuctShape(this Autodesk.Rev
...
+            if (connectorShape == Autodesk.Revit.DB.ConnectorProfileType.Rectangular)
                 return Autodesk.Revit.DB.ConnectorProfileType.Rectangular; // The duct is rectangular
-            else if (ductFamilyName.Contains("Round")) // Does the duct have a diameter?
-                return Autodesk.Revit.DB.ConnectorProfileType.Round; // The duct is circular
-            else if (ductFamilyName.Contains("Oval")) // Does the duct have a diameter?
-                return Autodesk.Revit.DB.ConnectorProfileType.Oval; // The duct is circular
+            else if (connectorShape == Autodesk.Revit.DB.ConnectorProfileType.Round)
+                return Autodesk.Revit.DB.ConnectorProfileType.Round; // The duct is round
+            else if (connectorShape == Autodesk.Revit.DB.ConnectorProfileType.Oval)
+                return Autodesk.Revit.DB.ConnectorProfileType.Oval; // The duct is oval
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/DuctShape.cs:71]: Why not `return connectorShape`?

File: Revit_Core_Engine/Query/OrientationAngle.cs [hunk 6]
```diff
@@ -58,31 +57,38 @@ public static double OrientationAngle(this FamilyInstance familyInstance, RevitS
         [Description("Get the orientation angle of a duct.")]
         [Input("Autodesk.Revit.DB.Mechanical.Duct", "Revit duct.")]
         [Input("BH.oM.Adapters.Revit.Settings.RevitSettings", "Revit settings.")]
-        [Output("double", "Orientation angle of a duct.")]
+        [Output("double", "Orientation angle of a duct in radians.")]
         public static double OrientationAngle(this Duct duct, RevitSettings settings = null)
         {
             settings = settings.DefaultIfNull();
+            Options options = new Options();
+            options.IncludeNonVisibleObjects = false;
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/OrientationAngle.cs:65]: What is it for?

File: Revit_Core_Engine/Query/OrientationAngle.cs [hunk 7]
```diff
@@ -58,31 +57,38 @@ public static double OrientationAngle(this FamilyInstance familyInstance, RevitS
...
-
             BH.oM.Geometry.ICurve bhomCurve = curve.IFromRevit(); // Convert to a BHoM curve
 
+            // Get the duct connector
+            Connector connector = null;
+            foreach (Connector conn in duct.ConnectorManager.Connectors)
+            {
+                connector = conn;
+                break;
+            }
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/OrientationAngle.cs:82]: As above, this should be an end connector, not any connector.

File: Revit_Core_Engine/Query/OrientationAngle.cs [hunk 8]
```diff
@@ -58,31 +57,38 @@ public static double OrientationAngle(this FamilyInstance familyInstance, RevitS
         [Description("Get the orientation angle of a duct.")]
         [Input("Autodesk.Revit.DB.Mechanical.Duct", "Revit duct.")]
         [Input("BH.oM.Adapters.Revit.Settings.RevitSettings", "Revit settings.")]
-        [Output("double", "Orientation angle of a duct.")]
+        [Output("double", "Orientation angle of a duct in radians.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/OrientationAngle.cs:60]: Variable names are wrong, descriptions could be improved too.

File: Revit_Core_Engine/Query/PipeSectionProperty.cs [hunk 9]
```diff
@@ -38,23 +38,23 @@ public static partial class Query
...
         /***************************************************/
 
-        [Description("Pipe section property converted fom Revit.")]
-        [Input("pipe", "Revit pipe type to be converted to a section property.")]
+        [Description("Convert a Revit pipe into a BHoM pipe section property.")]
+        [Input("pipe", "Revit pipe to be converted into a BHoM section property.")]
         [Input("settings", "Revit settings.")]
         [Input("refObjects", "Referenced objects.")]
-        [Output("property", "BHoM duct section property.")]
+        [Output("sectionProperty", "BHoM pipe section property converted from a Revit pipe.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/PipeSectionProperty.cs:45]: It is not conversion, rather query/extraction.

**Review by pawelbaran — CHANGES_REQUESTED** [review]
A few changes requested - please see the code comments.

Besides that, vertical ducts seem to have wrong orientation angle. The duct below, h > w comes out with rotation == 90 deg:

image

I have not reviewed the engineering value of the output as it is out of my expertise - @kayleighhoude @michaelhoehn I leave it for you. However, if the files are meant to become the standardized ones:
- there should be separation of concerns: geometry, section properties, possibly 3rd category being engineering properties (again, out of my field of expertise)
- duct geometry/location test file seems to be quite comprehensive, but e.g. pipe location does not even test sloped ones
- GH scripts are hard to follow: there is a lot of outputs, but e.g. there is no preview of the orientation

Again, please feel free to use the scripts from [here](https://burohappold.sharepoint.com/sites/BHoM/Installers/Forms/AllItems.aspx?FolderCTID=0x012000181C071E8B6D1E438B59C87DA36B9302&viewid=9079a548%2Dd015%2D48f3%2D842d%2D253576f13d8c&id=%2Fsites%2FBHoM%2FInstallers%2FScripts%2FBuroHappold%5FBHoM%5Fv3%2E3%2F0001%5FRevit%5FToolkit%2FPull) as inspiration.

File: Revit_Core_Engine/Query/PipeSectionProperty.cs (hunk 9 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/PipeSectionProperty.cs:45]: Revised accordingly

File: Revit_Core_Engine/Convert/Environment/FromRevit/Pipe.cs (hunk 1 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Convert/Environment/FromRevit/Pipe.cs:42]: All instances updated

File: Revit_Core_Engine/Query/OrientationAngle.cs (hunk 8 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/OrientationAngle.cs:60]: Corrected

File: Revit_Core_Engine/Query/OrientationAngle.cs (hunk 6 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/OrientationAngle.cs:65]: Boilerplate removed

File: Revit_Core_Engine/Query/DuctSectionProperty.cs (hunk 2 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/DuctSectionProperty.cs:43]: Revised

File: Revit_Core_Engine/Query/DuctSectionProfile.cs [hunk 10]
```diff
@@ -54,23 +54,14 @@ public static BH.oM.MEP.SectionProperties.SectionProfile DuctSectionProfile(this
             // Insulation thickness
             double insulationThickness = revitDuct.LookupParameterDouble(BuiltInParameter.RBS_INSULATION_THICKNESS);
```
Comment (michaelhoehn) [inline Revit_Core_Engine/Query/DuctSectionProfile.cs:55]: Should be changed to "RBS_REFERENCE_INSULATION_THICKNESS" as the current param value returns zero. This switch has worked on my tests to return accurate values for ducts and all profiles. More testing is required for pipes etc. 

Likewise, above for the lining "RBS_REFERENCE_LINING_THICKNESS"

**Review by michaelhoehn — CHANGES_REQUESTED** [review]

File: Revit_Core_Engine/Query/DuctSectionProfile.cs (hunk 3 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/DuctSectionProfile.cs:44]: Edited accordingly

File: Revit_Core_Engine/Query/DuctSectionProfile.cs (hunk 10 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/DuctSectionProfile.cs:55]: This is odd, the built in parameters are mapped to their corresponding string parameter names, but for some reason it doesn't return anything.
image

image

Comment (michaelhoehn) [reply Revit_Core_Engine/Query/DuctSectionProfile.cs:55]: Agreed, I was seeing the same thing.

File: Revit_Core_Engine/Query/DuctShape.cs (hunk 5 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/DuctShape.cs:71]: Good point, revised the code to return the connector shape

File: Revit_Core_Engine/Query/OrientationAngle.cs (hunk 7 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/OrientationAngle.cs:82]: Revised to find the right connector

File: Revit_Core_Engine/Query/DuctShape.cs (hunk 4 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/DuctShape.cs:54]: Revised to find the right connector

**Comment by AndreBalasian** [comment]
> * duct geometry/location test file seems to be quite comprehensive, but e.g. pipe location does not even test sloped ones

As discussed in the recent chat, this part is correct because we are not extracting the orientation angle from pipes as part of this PR.

**Comment by AndreBalasian** [comment]
> * GH scripts are hard to follow: there is a lot of outputs, but e.g. there is no preview of the orientation

As discussed in the recent chat, the orientation angle is only extracted for rectangular ducts.
All other test models are simplified.

**Comment by AndreBalasian** [comment]
> Besides that, vertical ducts seem to have wrong orientation angle. The duct below, h > w comes out with rotation == 90 deg:

Is this duct turned over on its side? If so, then 90 deg would be an accurate reading.

File: Revit_Core_Engine/Query/DuctSectionProfile.cs (hunk 10 above)
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/DuctSectionProfile.cs:55]: Fixed

File: Revit_Core_Engine/Query/DuctSectionProfile.cs [hunk 11]
```diff
@@ -36,41 +36,34 @@ public static partial class Query
...
         public static BH.oM.MEP.SectionProperties.SectionProfile DuctSectionProfile(this Autodesk.Revit.DB.Mechanical.Duct revitDuct, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
         {
             settings = settings.DefaultIfNull();
 
             IProfile profile = revitDuct.Profile(settings, refObjects);
 
             // Lining thickness
-            double liningThickness = revitDuct.LookupParameterDouble(BuiltInParameter.RBS_LINING_THICKNESS);
+            //double liningThickness = revitDuct.LookupParameterDouble("Lining Thickness");
+            double liningThickness = revitDuct.LookupParameterDouble(BuiltInParameter.RBS_REFERENCE_LINING_THICKNESS);
```
Comment (michaelhoehn) [inline Revit_Core_Engine/Query/DuctSectionProfile.cs:53]: This functions as expected and am happy with the functionality. Question for @pawelbaran prior to approval if this is an acceptable practice as this was my earlier suggestion to use the RBS_REFERENCE params.

File: Revit_Core_Engine/Query/DuctSectionProfile.cs [hunk 12]
```diff
@@ -36,41 +36,34 @@ public static partial class Query
...
 
             // Insulation thickness
-            double insulationThickness = revitDuct.LookupParameterDouble(BuiltInParameter.RBS_INSULATION_THICKNESS);
-
-            // Get the duct shape, which is either circular, rectangular, oval or null
-            Autodesk.Revit.DB.ConnectorProfileType ductShape = BH.Revit.Engine.Core.Query.DuctShape(revitDuct, settings);//revitDuct.DuctType.Shape;
+            //double insulationThickness = revitDuct.LookupParameterDouble("Insulation Thickness");
+            double insulationThickness = revitDuct.LookupParameterDouble(BuiltInParameter.RBS_REFERENCE_INSULATION_THICKNESS);
```
Comment (michaelhoehn) [inline Revit_Core_Engine/Query/DuctSectionProfile.cs:57]: Same as above

**Review by kayleighhoude — CHANGES_REQUESTED** [review]
Add ability to see 0 value on thickness for lining and insulation (rather than null) when the value does not exist

image

**Comment by michaelhoehn** [comment]
> Add ability to see 0 value on thickness for lining and insulation (rather than null) when the value does not exist

@kayleighhoude I'm seeing it's because the ShapeProfiles have different properties and can therefore not be represented in the same explode perhaps? Hope the image adds some clarification and possible solutions other than separating per shape profile. 

image

**Comment by AndreBalasian** [comment]
> Add ability to see 0 value on thickness for lining and insulation (rather than null) when the value does not exist
> 
> image

Added checks and made them return zeros from the Revit before being fed into the section profile create method.
However, in order for this to work, the zero checks on the section profile create methods in the engine would need to change.
image

**Comment by pawelbaran** [comment]
> > * duct geometry/location test file seems to be quite comprehensive, but e.g. pipe location does not even test sloped ones
> 
> As discussed in the recent chat, this part is correct because we are not extracting the orientation angle from pipes as part of this PR.

As discussed in the recent chat, orientation angle and slope are two separate concepts. I am asking for slope, not orientation.

**Comment by pawelbaran** [comment]
> > * GH scripts are hard to follow: there is a lot of outputs, but e.g. there is no preview of the orientation
> 
> As discussed in the recent chat, the orientation angle is only extracted for rectangular ducts.
> All other test models are simplified.

Yes, but without a preview of pulled objects it is not easy to validate whether the sign of orientation angle value is correct or not - by creating preview of the elements, we can easily make sure that they are correct.

**Comment by FraserGreenroyd** [comment]
> Add ability to see 0 value on thickness for lining and insulation (rather than null) when the value does not exist

If it's 0, it exists, but is highly invalid. If it's null, it doesn't. I would agree with @AndreBalasian on this one, we should keep it to `null` when it doesn't exist.

**Comment by pawelbaran** [comment]
> > Besides that, vertical ducts seem to have wrong orientation angle. The duct below, h > w comes out with rotation == 90 deg:
> 
> Is this duct turned over on its side? If so, then 90 deg would be an accurate reading.

It is a top view - vertical ducts are rotated by 90 deg.

**Comment by AndreBalasian** [comment]
> > > * duct geometry/location test file seems to be quite comprehensive, but e.g. pipe location does not even test sloped ones
> > 
> > 
> > As discussed in the recent chat, this part is correct because we are not extracting the orientation angle from pipes as part of this PR.
> 
> As discussed in the recent chat, orientation angle and slope are two separate concepts. I am asking for slope, not orientation.

Slope will not being extracted in this PR and therefore no testing is needed for slopes

File: Revit_Core_Engine/Query/DuctSectionProfile.cs (hunk 11 above)
Comment (pawelbaran) [reply Revit_Core_Engine/Query/DuctSectionProfile.cs:52]: Yeah, I think it is fine, as long as the correct parameter value is extracted - if you know which parameter value you want to extract from Revit, you can check its underlying `BuiltinParameter` value using Revit Lookup. That should be the ultimate answer to your question 😉 

By the way, could you please make sure that no commented code has been left behind @AndreBalasian?

**Comment by pawelbaran** [comment]
> > > > * duct geometry/location test file seems to be quite comprehensive, but e.g. pipe location does not even test sloped ones
> > > 
> > > 
> > > As discussed in the recent chat, this part is correct because we are not extracting the orientation angle from pipes as part of this PR.
> > 
> > 
> > As discussed in the recent chat, orientation angle and slope are two separate concepts. I am asking for slope, not orientation.
> 
> Slope will not being extracted in this PR and therefore no testing is needed for slopes

Coming back to my original statement: 
> duct geometry/location** test file seems to be quite comprehensive, but e.g. pipe location does not even test sloped ones.

By **geometry/location** I mean the location of a pipe in 3-dimensional space. The test file attached to this PR contains only horizontal pipes, all of them in one plane. What we also need to test are other configurations: vertical pipe, sloped pipe, pipe with justification other than center etc. Same applies to other types (duct, wire) too. This one has nothing to do with numerical values, it is about the location of the start node and end node - orientation is a separate subject.

**Comment by pawelbaran** [comment]
> > > Besides that, vertical ducts seem to have wrong orientation angle. The duct below, h > w comes out with rotation == 90 deg:
> > 
> > 
> > Is this duct turned over on its side? If so, then 90 deg would be an accurate reading.
> 
> It is a top view - vertical ducts are rotated by 90 deg.

Double checked upon @AndreBalasian's request - my mistake, the current behaviour is correct. 👍

**Comment by AndreBalasian** [comment]
> > > > > * duct geometry/location test file seems to be quite comprehensive, but e.g. pipe location does not even test sloped ones
> > > > 
> > > > 
> > > > As discussed in the recent chat, this part is correct because we are not extracting the orientation angle from pipes as part of this PR.
> > > 
> > > 
> > > As discussed in the recent chat, orientation angle and slope are two separate concepts. I am asking for slope, not orientation.
> > 
> > 
> > Slope will not being extracted in this PR and therefore no testing is needed for slopes
> 
> Coming back to my original statement:
> 
> > duct geometry/location** test file seems to be quite comprehensive, but e.g. pipe location does not even test sloped ones.
> 
> By **geometry/location** I mean the location of a pipe in 3-dimensional space. The test file attached to this PR contains only horizontal pipes, all of them in one plane. What we also need to test are other configurations: vertical pipe, sloped pipe, pipe with justification other than center etc. Same applies to other types (duct, wire) too. This one has nothing to do with numerical values, it is about the location of the start node and end node - orientation is a separate subject.

Made the round ducts and pipe files more comprehensive as requested. Wire lines to remain in the 2D floor plan due to limitations with Revit's wire visualisation.

**Comment by AndreBalasian** [comment]
> > > * GH scripts are hard to follow: there is a lot of outputs, but e.g. there is no preview of the orientation
> > 
> > 
> > As discussed in the recent chat, the orientation angle is only extracted for rectangular ducts.
> > All other test models are simplified.
> 
> Yes, but without a preview of pulled objects it is not easy to validate whether the sign of orientation angle value is correct or not - by creating preview of the elements, we can easily make sure that they are correct.

Previews added to the test script as requested.
For some reason all I can see are the start and end points of objects.

File: Revit_Core_Engine/Query/DuctSectionProfile.cs (hunk 11 above)
Comment (michaelhoehn) [reply Revit_Core_Engine/Query/DuctSectionProfile.cs:52]: Thanks @pawelbaran and @AndreBalasian seems to be functioning as expected with the latest changes. I'm happy to consider my specific query resolved. Please mark resolved once the commented code has been removed as requested. 😄

**Review by michaelhoehn — DISMISSED** [review]

**Comment by michaelhoehn** [comment]
Approving functionality. I've successfully tested and am happy with the functionality.

**Review by FraserGreenroyd — DISMISSED** [review]
Based on my discussions with the team, I am happy with this being merged when the following is also true:

 - @michaelhoehn is happy and approves from a functionality/discipline side
 - @kayleighhoude is happy and approves from a functionality/discipline side
 - @pawelbaran is happy and approves from a code framework side

But it's looking promising from my end :smile:

**Review by michaelhoehn — DISMISSED** [review]
Approved based on functionality. Happy with the test results and code descriptions.

**Review by kayleighhoude — DISMISSED** [review]
Had to give it one last test!

**Review by pawelbaran — CHANGES_REQUESTED** [review]
I took another round through the code, in general happy with the looks of it, only a few cosmetic changes requested (they should not change the mechanics of the code). Regarding the comment on `OrientationAngle` I am happy to approve and merge the code in its current state, but would be good to further investigate that subject while testing the Beta and fix then if needed.

File: Revit_Core_Engine/Query/DuctSectionProfile.cs [hunk 13]
```diff
@@ -36,41 +36,36 @@ public static partial class Query
...
         /***************************************************/
-
-        [Description("Section profile of round and rectangular ducts converted from a Revit duct.")]
-        [Input("revitDuct", "Revit duct to be converted into a section profile.")]
-        [Input("settings", "Revit settings.")]
-        [Input("refObjects", "Referenced objects.")]
-        [Output("profile", "Converted BHoM duct section profile converted.")]
+        
+        [Description("Query a Revit duct to extract a BHoM duct section profile.")]
+        [Input("revitDuct", "Revit duct to be converted into a BHoM section profile.")]
```
[the remark this answers is not in the document]
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/DuctSectionProfile.cs:41]: Revised accordingly

File: Revit_Core_Engine/Query/DuctSectionProfile.cs [hunk 14]
```diff
@@ -36,41 +36,36 @@ public static partial class Query
...
+        [Input("settings", "Revit adapter settings.")]
+        [Input("refObjects", "A collection of objects processed in the current adapter action, stored to avoid processing the same object more than once.")]
+        [Output("profile", "BHoM section profile for a duct extracted from a Revit duct.")]
         public static BH.oM.MEP.SectionProperties.SectionProfile DuctSectionProfile(this Autodesk.Revit.DB.Mechanical.Duct revitDuct, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
```
[the remark this answers is not in the document]
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/DuctSectionProfile.cs:45]: refObjects removed

File: Revit_Core_Engine/Query/DuctSectionProperty.cs [hunk 15]
```diff
@@ -40,21 +36,19 @@ public static partial class Query
         /****               Public Methods              ****/
         /***************************************************/
 
-        [Description("Duct section property converted from a Revit duct.")]
-        [Input("revitDuct", "Revit duct to be conerted into a section property.")]
-        [Input("settings", "Revit settings.")]
-        [Input("refObjects", "Referenced objects.")]
-        [Output("DuctSectionProperty", "BHoM duct section property.")]
+        [Description("Query a Revit duct to extract a BHoM duct section property.")]
+        [Input("revitDuct", "Revit duct to be converted into a BHoM section property.")]
```
[the remark this answers is not in the document]
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/DuctSectionProperty.cs:40]: Revised

File: Revit_Core_Engine/Query/DuctSectionProperty.cs [hunk 16]
```diff
@@ -40,21 +36,19 @@ public static partial class Query
...
+        [Input("settings", "Revit adapter settings.")]
+        [Input("refObjects", "A collection of objects processed in the current adapter action, stored to avoid processing the same object more than once.")]
+        [Output("sectionProperty", "BHoM duct section property extracted from a Revit duct.")]
         public static BH.oM.MEP.SectionProperties.DuctSectionProperty DuctSectionProperty(this Autodesk.Revit.DB.Mechanical.Duct revitDuct, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
```
[the remark this answers is not in the document]
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/DuctSectionProperty.cs:44]: refObjects removed

File: Revit_Core_Engine/Query/PipeSectionProperty.cs [hunk 17]
```diff
@@ -38,25 +38,25 @@ public static partial class Query
         /****               Public Methods              ****/
         /***************************************************/
 
-        [Description("Pipe section property converted fom Revit.")]
-        [Input("pipe", "Revit pipe type to be converted to a section property.")]
-        [Input("settings", "Revit settings.")]
-        [Input("refObjects", "Referenced objects.")]
-        [Output("property", "BHoM duct section property.")]
+        [Description("Query a Revit pipe to get a BHoM pipe section property.")]
+        [Input("pipe", "Revit pipe to be converted into a BHoM section property.")]
```
[the remark this answers is not in the document]
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/PipeSectionProperty.cs:42]: Revised

File: Revit_Core_Engine/Query/PipeSectionProperty.cs [hunk 18]
```diff
@@ -38,25 +38,25 @@ public static partial class Query
...
+        [Input("settings", "Revit adapter settings.")]
+        [Input("refObjects", "A collection of objects processed in the current adapter action, stored to avoid processing the same object more than once.")]
+        [Output("sectionProperty", "BHoM pipe section property extracted from a Revit pipe.")]
         public static BH.oM.MEP.SectionProperties.PipeSectionProperty PipeSectionProperty(this Autodesk.Revit.DB.Plumbing.Pipe pipe, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
```
[the remark this answers is not in the document]
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/PipeSectionProperty.cs:46]: refObjects removed

File: Revit_Core_Engine/Query/Profile.cs [hunk 19]
```diff
@@ -39,17 +38,15 @@ public static partial class Query
...
-        [Description("Extract a BHoM profile from a Revit duct.")]
-        [Input("duct", "Revit duct for property extraction.")]
-        [Input("settings", "Revit settings.")]
-        [Input("refObjects", "Referenced objects.")]
+        [Description("Extract a BHoM duct profile from a Revit duct.")]
+        [Input("duct", "Revit duct to extract relevant profie information from.")]
+        [Input("settings", "Revit adapter settings.")]
+        [Input("refObjects", "A collection of objects processed in the current adapter action, stored to avoid processing the same object more than once.")]
         [Output("profile", "BHoM duct profile extracted from a Revit duct.")]
         public static IProfile Profile(this Autodesk.Revit.DB.Mechanical.Duct duct, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
```
[the remark this answers is not in the document]
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/Profile.cs:46]: refObjects removed

File: Revit_Core_Engine/Query/Profile.cs [hunk 20]
```diff
@@ -84,10 +82,10 @@ public static IProfile Profile(this Autodesk.Revit.DB.Mechanical.Duct duct, Revi
...
-        [Description("Extract a BHoM profile from a Revit pipe.")]
-        [Input("pipe", "Revit pipe for property extraction.")]
-        [Input("settings", "Revit settings.")]
-        [Input("refObjects", "Referenced objects.")]
+        [Description("Extract a BHoM pipe profile from a Revit pipe.")]
+        [Input("pipe", "Revit pipe to extract relevant profie information from.")]
+        [Input("settings", "Revit adapter settings.")]
+        [Input("refObjects", "A collection of objects processed in the current adapter action, stored to avoid processing the same object more than once.")]
         [Output("profile", "BHoM pipe profile extracted from a Revit pipe.")]
         public static IProfile Profile(this Autodesk.Revit.DB.Plumbing.Pipe pipe, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
```
[the remark this answers is not in the document]
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/Profile.cs:90]: refObjects removed

File: Revit_Core_Engine/Query/Profile.cs [hunk 21]
```diff
@@ -102,16 +100,16 @@ public static IProfile Profile(this Autodesk.Revit.DB.Plumbing.Pipe pipe, RevitS
...
-        [Input("wire", "Revit wire for property extraction.")]
-        [Input("settings", "Revit settings.")]
-        [Input("refObjects", "Referenced objects.")]
-        [Output("profile", "BHoM wire profile extracted from a Revit wire.")]
+        [Description("Extract a BHoM tube profile from a Revit wire. A tube profile is used in this case to facilitate the extraction of the wire diameter because a dedicated BHoM profile for a wire is not available.")]
+        [Input("wire", "Revit wire to extract relevant profie information from.")]
+        [Input("settings", "Revit adapter settings.")]
+        [Input("refObjects", "A collection of objects processed in the current adapter action, stored to avoid processing the same object more than once.")]
+        [Output("profile", "BHoM tube profile extracted from a Revit wire.")]
         public static IProfile Profile(this Autodesk.Revit.DB.Electrical.Wire wire, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
```
[the remark this answers is not in the document]
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/Profile.cs:113]: refObjects removed

File: Revit_Core_Engine/Query/OrientationAngle.cs [hunk 22]
```diff
@@ -55,34 +54,43 @@ public static double OrientationAngle(this FamilyInstance familyInstance, RevitS
...
-            XYZ endPoint = curve.GetEndPoint(1); // End point
-
-            XYZ vector = (endPoint - startPoint).Normalize(); // Normalised vector
+            BH.oM.Geometry.ICurve bhomCurve = curve.IFromRevit(); // Convert to a BHoM curve
 
-            Transform transform = Transform.CreateTranslation(vector); // Transform of a vector
+            // Get the duct connector
+            Connector connector = null;
+            foreach (Connector conn in duct.ConnectorManager.Connectors)
+            {
```
[the remark this answers is not in the document]
Comment (AndreBalasian) [reply Revit_Core_Engine/Query/OrientationAngle.cs:77]: Comment noted

**Review by pawelbaran — APPROVED** [review]
Code LGTM

**Review by FraserGreenroyd — DISMISSED** [review]

**Comment by pawelbaran** [comment]
@michaelhoehn @kayleighhoude there should not be any functional changes since your last review, but just to double check - are you happy with merging the PR?

**Comment by michaelhoehn** [comment]
@pawelbaran LGTM! Thanks 😄

**Review by FraserGreenroyd — APPROVED** [review]
Approving following rebase, no code changes since last review.

**Review by michaelhoehn — APPROVED** [review]
LGTM
